File: lib/start-lock.py
"""Serialize flow-start with a queue directory.

Prevents concurrent starts from fighting over main (CI fixes, dependency
updates). Only one flow-start runs at a time. The oldest queue entry
(by mtime, then feature name) holds the lock.

Usage:
    bin/flow start-lock --acquire --feature <name>
    bin/flow start-lock --acquire --wait --feature <name>
    bin/flow start-lock --acquire --wait --timeout 90 --feature <name>
    bin/flow start-lock --release --feature <name>
    bin/flow start-lock --check

Output (JSON to stdout, all responses include "lock_path"):
    Acquire: {"status": "acquired", "lock_path": ...}
             or {"status": "locked", "feature": ..., "lock_path": ...}
    Acquire --wait: {"status": "acquired", ...}
             or {"status": "timeout", ..., "lock_path": ...}
    Release: {"status": "released", "lock_path": ...}
             or {"status": "error", "message": ..., "lock_path": ...}
    Check:   {"status": "free", "lock_path": ...}
             or {"status": "locked", ..., "lock_path": ...}
"""

import argparse
import json
import sys
import time
from pathlib import Path

sys.path.insert(0, str(Path(__file__).resolve().parent))

from flow_utils import project_root

QUEUE_DIRNAME = "start-queue"
STALE_TIMEOUT_SECONDS = 1800  # 30 minutes

_CACHED_QUEUE_PATH = None


def _queue_path():
    """Return the path to the queue directory. Cached after first resolution."""
    global _CACHED_QUEUE_PATH
    if _CACHED_QUEUE_PATH is not None:
        return _CACHED_QUEUE_PATH
    root = project_root().resolve()
    state_dir = root / ".flow-states"
    state_dir.mkdir(parents=True, exist_ok=True)
    queue_dir = state_dir / QUEUE_DIRNAME
    queue_dir.mkdir(exist_ok=True)
    _CACHED_QUEUE_PATH = queue_dir
    return _CACHED_QUEUE_PATH


def _list_queue(queue_dir, cleanup=False):
    """List queue entries sorted by (mtime, name).

    If cleanup=True, remove stale entries (>30 min) before sorting.
    Returns list of (mtime, name) tuples and whether any stale entries
    were removed.
    """
    stale_removed = False
    entries = []

    try:
        items = list(queue_dir.iterdir())
    except OSError:
        return [], False

    for item in items:
        if not item.is_file():
            continue
        try:
            mtime = item.stat().st_mtime
        except OSError:
            continue
        if cleanup and (time.time() - mtime) > STALE_TIMEOUT_SECONDS:
            item.unlink(missing_ok=True)
            stale_removed = True
            continue
        entries.append((mtime, item.name))

    entries.sort()
    return entries, stale_removed


def acquire(feature):
    """Attempt to acquire the start lock via the queue."""
    queue_dir = _queue_path()

    # Create our queue entry (idempotent — overwrites if exists)
    entry = queue_dir / feature
    entry.touch(exist_ok=True)

    # List queue with stale cleanup
    entries, stale_removed = _list_queue(queue_dir, cleanup=True)

    if not entries:
        # Should not happen — we just created our entry
        return {"status": "acquired", "lock_path": str(queue_dir)}

    holder = entries[0][1]
    if holder == feature:
        result = {"status": "acquired", "lock_path": str(queue_dir)}
        if stale_removed:
            result["stale_broken"] = True
        return result

    result = {
        "status": "locked",
        "feature": holder,
        "lock_path": str(queue_dir),
    }
    if stale_removed:
        result["stale_broken"] = True
    return result


def acquire_with_wait(feature, timeout=90, interval=10):
    """Acquire the lock, retrying with sleep until acquired or timed out."""
    start = time.monotonic()
    result = acquire(feature)
    if result["status"] == "acquired":
        return result

    while True:
        elapsed = time.monotonic() - start
        if elapsed >= timeout:
            return {
                "status": "timeout",
                "feature": result["feature"],
                "waited_seconds": int(elapsed),
                "lock_path": result["lock_path"],
            }
        remaining = timeout - elapsed
        time.sleep(min(interval, remaining))
        result = acquire(feature)
        if result["status"] == "acquired":
            return result


def release(feature):
    """Release the start lock by removing our queue entry."""
    queue_dir = _queue_path()
    entry = queue_dir / feature
    entry.unlink(missing_ok=True)
    if entry.exists():
        return {"status": "error", "message": "Queue entry persists after unlink", "lock_path": str(queue_dir)}
    return {"status": "released", "lock_path": str(queue_dir)}


def check():
    """Check the current lock status without modifying."""
    queue_dir = _queue_path()

    # Read-only check — filter stale but don't delete
    entries = []
    try:
        for item in queue_dir.iterdir():
            if not item.is_file():
                continue
            try:
                mtime = item.stat().st_mtime
            except OSError:
                continue
            if (time.time() - mtime) > STALE_TIMEOUT_SECONDS:
                continue
            entries.append((mtime, item.name))
    except OSError:
        pass

    if not entries:
        return {"status": "free", "lock_path": str(queue_dir)}

    entries.sort()
    holder = entries[0][1]
    return {
        "status": "locked",
        "feature": holder,
        "lock_path": str(queue_dir),
    }


def main():
    parser = argparse.ArgumentParser(description="FLOW start lock")
    parser.add_argument("--acquire", action="store_true", help="Acquire the lock")
    parser.add_argument("--release", action="store_true", help="Release the lock")
    parser.add_argument("--check", action="store_true", help="Check lock status")
    parser.add_argument("--feature", default=None, help="Feature name (required for --acquire and --release)")
    parser.add_argument("--wait", action="store_true", help="Wait for lock to be released")
    parser.add_argument("--timeout", type=int, default=90, help="Max seconds to wait (default 90)")
    args = parser.parse_args()

    if args.acquire:
        if not args.feature:
            print(json.dumps({"status": "error", "message": "--feature required for --acquire"}))
            sys.exit(1)
        if args.wait:
            result = acquire_with_wait(args.feature, timeout=args.timeout)
        else:
            result = acquire(args.feature)
    elif args.release:
        if not args.feature:
            print(json.dumps({"status": "error", "message": "--feature required for --release"}))
            sys.exit(1)
        result = release(args.feature)
    elif args.check:
        result = check()
    else:
        print(json.dumps({"status": "error", "message": "Specify --acquire, --release, or --check"}))
        sys.exit(1)

    print(json.dumps(result))


if __name__ == "__main__":
    main()

File: src/commands/start_lock.rs
//! Queue-based start lock serialization.
//!
//! Prevents concurrent starts from fighting over main (CI fixes, dependency
//! updates). Only one flow-start runs at a time. The oldest queue entry
//! (by mtime, then feature name) holds the lock.

use std::fs;
use std::path::{Path, PathBuf};
use std::time::{Duration, SystemTime, UNIX_EPOCH};

use serde_json::{json, Value};

use crate::flow_paths::FlowStatesDir;

pub const QUEUE_DIRNAME: &str = "start-queue";
pub const STALE_TIMEOUT_SECONDS: u64 = 1800; // 30 minutes

/// Get file mtime as seconds since UNIX epoch. Returns `0.0` on any
/// failure (metadata error, mtime unsupported, or a pre-UNIX_EPOCH
/// mtime). Callers treat `0.0` as maximally stale, so a broken entry
/// is classified as stale and cleaned up on the next list_queue pass
/// — the same outcome we'd get from an explicit skip branch, but
/// expressed without a fallible return type.
fn mtime_secs(path: &Path) -> f64 {
    fs::metadata(path)
        .and_then(|m| m.modified())
        .map(|t| {
            t.duration_since(UNIX_EPOCH)
                .map(|d| d.as_secs_f64())
                .unwrap_or(0.0)
        })
        .unwrap_or(0.0)
}

/// Create the queue directory if needed, return its path.
pub fn queue_path(root: &Path) -> PathBuf {
    let root = match root.canonicalize() {
        Ok(p) => p,
        Err(_) => root.to_path_buf(),
    };
    // The queue lives under `.flow-states/` and is shared across every
    // branch on this machine, so FlowStatesDir (branch-free) is the
    // right address for it.
    let state_dir = FlowStatesDir::new(&root).path().to_path_buf();
    let _ = fs::create_dir_all(&state_dir);
    let queue_dir = state_dir.join(QUEUE_DIRNAME);
    let _ = fs::create_dir_all(&queue_dir);
    queue_dir
}

/// List queue entries sorted by (mtime_secs, name).
///
/// Stale entries (older than STALE_TIMEOUT_SECONDS) are always excluded
/// from the returned list. If `cleanup` is true, they are also deleted.
/// Returns (entries, stale_found) where stale_found indicates whether
/// any stale entries were encountered.
pub fn list_queue(queue_dir: &Path, cleanup: bool) -> (Vec<(f64, String)>, bool) {
    let mut stale_found = false;
    let mut entries: Vec<(f64, String)> = Vec::new();

    let items = match fs::read_dir(queue_dir) {
        Ok(items) => items,
        Err(_) => return (entries, false),
    };

    let now_secs = SystemTime::now()
        .duration_since(UNIX_EPOCH)
        .unwrap_or_default()
        .as_secs_f64();

    for entry in items.flatten() {
        let path = entry.path();
        // Skip non-files (directories, symlinks, etc.)
        if !path.is_file() {
            continue;
        }
        let mtime = mtime_secs(&path);
        if (now_secs - mtime) > STALE_TIMEOUT_SECONDS as f64 {
            if cleanup {
                let _ = fs::remove_file(&path);
            }
            stale_found = true;
            continue; // stale entries excluded from list
        }
        let name = entry.file_name().to_string_lossy().to_string();
        entries.push((mtime, name));
    }

    entries.sort_by(|a, b| {
        a.0.partial_cmp(&b.0)
            .unwrap_or(std::cmp::Ordering::Equal)
            .then_with(|| a.1.cmp(&b.1))
    });

    (entries, stale_found)
}

/// Attempt to acquire the start lock via the queue.
///
/// Creates a queue entry for this feature. If we are first in queue
/// (by mtime, then name), returns acquired. Otherwise returns locked
/// with the current holder's name.
pub fn acquire(feature: &str, queue_dir: &Path) -> Value {
    let lock_path = queue_dir.display().to_string();
    let entry = queue_dir.join(feature);

    // Create our queue entry only if it doesn't exist yet.
    // If an existing entry is stale (from a previous incomplete run),
    // replace it so _list_queue cleanup doesn't delete our only entry.
    if entry.exists() {
        let mtime = mtime_secs(&entry);
        let now_secs = SystemTime::now()
            .duration_since(UNIX_EPOCH)
            .unwrap_or_default()
            .as_secs_f64();
        if (now_secs - mtime) > STALE_TIMEOUT_SECONDS as f64 {
            let _ = fs::remove_file(&entry);
            let _ = fs::File::create(&entry);
        }
    } else {
        let _ = fs::File::create(&entry);
    }

    // List queue with stale cleanup
    let (entries, stale_removed) = list_queue(queue_dir, true);

    if entries.is_empty() {
        // Should not happen — we just created our entry
        let mut result = json!({"status": "acquired", "lock_path": lock_path});
        if stale_removed {
            result["stale_broken"] = json!(true);
        }
        return result;
    }

    let holder = &entries[0].1;
    if holder == feature {
        let mut result = json!({"status": "acquired", "lock_path": lock_path});
        if stale_removed {
            result["stale_broken"] = json!(true);
        }
        result
    } else {
        let mut result = json!({
            "status": "locked",
            "feature": holder.clone(),
            "lock_path": lock_path,
        });
        if stale_removed {
            result["stale_broken"] = json!(true);
        }
        result
    }
}

/// Acquire with retry loop using the real thread::sleep.
pub fn acquire_with_wait(feature: &str, queue_dir: &Path, timeout: u64, interval: u64) -> Value {
    acquire_with_wait_impl(
        feature,
        queue_dir,
        timeout,
        interval,
        &mut |d| std::thread::sleep(d),
    )
}

/// Seam-injected variant of [`acquire_with_wait`] that accepts a
/// custom sleep closure. Tests substitute a no-op or side-effecting
/// closure to drive every branch without blocking on real time.
///
/// Takes `&mut dyn FnMut(Duration)` rather than a generic `F`
/// parameter so every caller's closure compiles into the SAME
/// monomorphization. Generics create one monomorphization per
/// caller type, and callers from other test binaries (or from
/// code paths that never execute in the per-file test binary)
/// show up as `Unexecuted instantiation`, inflating the
/// uncovered-region/line counts. See
/// `.claude/rules/rust-patterns.md` "Seam-injection variant for
/// externally-coupled code" and the coverage-quest prompt's
/// "Generic functions" note.
pub fn acquire_with_wait_impl(
    feature: &str,
    queue_dir: &Path,
    timeout: u64,
    interval: u64,
    sleep_fn: &mut dyn FnMut(Duration),
) -> Value {
    let start = std::time::Instant::now();
    let result = acquire(feature, queue_dir);
    if result["status"] == "acquired" {
        return result;
    }

    loop {
        let elapsed = start.elapsed().as_secs();
        if elapsed >= timeout {
            return json!({
                "status": "timeout",
                "feature": result["feature"],
                "waited_seconds": elapsed as i64,
                "lock_path": result["lock_path"],
            });
        }
        let remaining = timeout - elapsed;
        sleep_fn(Duration::from_secs(std::cmp::min(interval, remaining)));
        let result = acquire(feature, queue_dir);
        if result["status"] == "acquired" {
            return result;
        }
    }
}

/// Release the start lock by removing the queue entry.
///
/// Returns `was_present: true` if the file existed before removal,
/// `false` if it was already absent. Status is `"released"` in both
/// cases (idempotent). `"error"` only if the file persists after unlink.
pub fn release(feature: &str, queue_dir: &Path) -> Value {
    let lock_path = queue_dir.display().to_string();
    let entry = queue_dir.join(feature);
    let was_present = entry.exists();
    let _ = fs::remove_file(&entry);

    if entry.exists() {
        return json!({
            "status": "error",
            "message": "Queue entry persists after unlink",
            "lock_path": lock_path,
            "was_present": true,
        });
    }

    json!({"status": "released", "lock_path": lock_path, "was_present": was_present})
}

/// Check lock status without modifying.
pub fn check(queue_dir: &Path) -> Value {
    let lock_path = queue_dir.display().to_string();
    let (entries, _) = list_queue(queue_dir, false);

    if entries.is_empty() {
        return json!({"status": "free", "lock_path": lock_path});
    }

    let holder = &entries[0].1;
    json!({
        "status": "locked",
        "feature": holder.clone(),
        "lock_path": lock_path,
    })
}

/// Testable core of the start-lock CLI. Returns the JSON payload the
/// CLI wrapper would print plus the exit code. The wrapper in
/// `main.rs` computes `project_root()` once and passes it in.
pub fn run_impl_main(
    acquire_flag: bool,
    release_flag: bool,
    check_flag: bool,
    feature: Option<String>,
    wait: bool,
    timeout: u64,
    interval: u64,
    root: &Path,
) -> (Value, i32) {
    if acquire_flag {
        let feature = match feature {
            Some(f) => f,
            None => {
                return (
                    json!({"status": "error", "message": "--feature required for --acquire"}),
                    1,
                )
            }
        };
        let queue_dir = queue_path(root);
        let result = if wait {
            acquire_with_wait(&feature, &queue_dir, timeout, interval)
        } else {
            acquire(&feature, &queue_dir)
        };
        (result, 0)
    } else if release_flag {
        let feature = match feature {
            Some(f) => f,
            None => {
                return (
                    json!({"status": "error", "message": "--feature required for --release"}),
                    1,
                )
            }
        };
        let queue_dir = queue_path(root);
        (release(&feature, &queue_dir), 0)
    } else if check_flag {
        let queue_dir = queue_path(root);
        (check(&queue_dir), 0)
    } else {
        (
            json!({"status": "error", "message": "Specify --acquire, --release, or --check"}),
            1,
        )
    }
}
